### PR TITLE
ENH Add configuration to cap report counts for large datasets.

### LINKED ISSUE #105: Cap the count of elements being counted in a content block report so it doesn't time out
Reports of Content Blocks currently time out when there are thousands of items to count within a report. We should limit the counter at an amount which wont cause the report to time out. 

The counter is displayed as a number, cap the number at 1000 (just example) and signify there are more...  eg. 1000+

This is a issue which has been raised by a site which uses over 7000 content blocks.

Related issue https://github.com/dnadesign/silverstripe-elemental/issues/238

### PR DESCRIPTION
**Pull request by GuySartorelli**
Fixes #105

Opted not to return the result from `getCount()` directly since the PHPDoc says this returns an int (even though it technically already returns a string if the report isn't set up correctly...)

### CODE REVIEW

File: code/Report.php [hunk 1]
```diff
@@ -218,16 +219,43 @@ protected function sanitiseClassName($class)
      * @param array $params - any parameters for the sourceRecords
      * @return int
      */
-    public function getCount($params = array())
+    public function getCount($params = array(), $limit = null)
```
Comment (emteknetnz) [inline code/Report.php:222]: I've just double checked that `$limit = null` is correct, as opposed to something like `int $limit = 0`

https://github.com/silverstripe/silverstripe-framework/blob/4/src/ORM/Queries/SQLSelect.php#L256

tl;dr - null is the correct way to set no limit and there are no params type hints for `$limit` through the ORM as it was written pre php7

**Review by emteknetnz — CHANGES_REQUESTED** [review]
I was a little skeptical at first of adding an optional `$limit` param to `getCount()`, though I think the way you have it here is correct

Would you be able to make requested changes + add a unit test for `getCountForOverview()`?

After that I'm happy to merge

File: README.md [hunk 2]
```diff
@@ -19,6 +19,14 @@ The reports section will not show up in the CMS if:
 
  * There are no reports to show
  * The logged in user does not have permission to view any reports
+ 
+For large datasets, the reports section may take a long time to load, since each report is getting a count of the items it contains to display next to the title.  
+You can mitigate this issue by setting a cap on the number of items that will be counted per report with the `limit_count_in_overview` configuration variable. You can either do for all reports, or if you know specifically which reports will be problematic you can set it for those specific reports.
```
[the remark this answers is not in the document]
Comment (GuySartorelli) [reply README.md:24]: This is a bit ambiguous because of how github has chosen to show the excerpt, so to clarify I take it you mean the paragraph that starts "You can mitigate"?

**Comment by GuySartorelli** [comment]
@emteknetnz Will make those changes as time permits - hopefully in the next week.

**Comment by sunnysideup** [comment]
will this get merged?

**Comment by sunnysideup** [comment]
Hopefully it can be added. It is a good one.

File: code/Report.php [hunk 3]
```diff
@@ -218,16 +219,43 @@ protected function sanitiseClassName($class)
...
+     * Large datasets can cause performance issues for some reports if allowed to count all records.
+     * To mitigate this, you can set the limit_count_in_overview config variable to the maximum number
+     * of items you wish to count to. Counts will be limited to this value, and any counts that hit
+     * this limit will be displayed with a plus, e.g. "500+"
+     *
+     * The default is to have no limit.
+     *
+     * @return string|int
+     */
+    public function getCountForOverview()
```
[the remark this answers is not in the document]
Comment (GuySartorelli) [reply code/Report.php:249]: I'll also update the PHPDoc to match.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (GuySartorelli) [reply README.md:24]: There was a line break before that paragraph using `  ` (two spaces) syntax - I'm changing this to a `\n` character for the sake of clarity.

**Comment by sunnysideup** [comment]
I's set the default limit to 10,000. I think it is good to have default limits on things. People can always increase it and 

10,000+ and 1,000,000+ is basically the same thing for a user -> MANY!

**Comment by GuySartorelli** [comment]
@emteknetnz What do you think about @sunnysideup's suggestion? Happy to make the change if it's approved.

**Comment by GuySartorelli** [comment]
Requested changes to date have been made, and tests are passing.

**Comment by emteknetnz** [comment]
Yup 10,000 default limit makes sense - presumably you'd add just add `private static $limit_count_in_overview = 10000;` on Report?

**Comment by GuySartorelli** [comment]
Yeah and I'll update the test to match. Sounds like a plan.

**Comment by GuySartorelli** [comment]
@emteknetnz I've added the default and updated the readme and test.

**Review by emteknetnz — APPROVED** [review]
